source/FormationPattern.cpp

```
@@ -91,9 +195,227 @@ void FormationPattern::SetName(const std::string &name)
+			}
+		}
+
+		// Apply positions to get the correct position-angle.
+		if(positions > 1)
+			endAngle /= positions - 1;
+		double positionAngle = startAngle + endAngle * linePosition;
+
+		// Get into the range of 0 to 360 for conversion to angle)
+		if(positionAngle < 0)
+			positionAngle = -fmod(-positionAngle, 360) + 360;
+		else
+			positionAngle = fmod(positionAngle, 360);
+
+		// Combine anchor with the position and return the result.
+		return endOrAnchorPx + Angle(positionAngle).Unit() * radius;
+	}
+
+	// This is not an arc, perform the line-based calculations.
+
+	// Swap start and end if we need to alternate in the repeat section.
+	// Apply repeat section if it is relevant.
+	if(ring % 2 && repeat && repeat->alternating)
+	{
+		Point tmpPx = endOrAnchorPx;
+		endOrAnchorPx = startPx;
+		startPx = tmpPx;
+	}
+
+	// Calculate the step from each position between start and end.
+	Point positionPx = endOrAnchorPx - startPx;
+
+	// Divide by positions, but don't count the first (since it is at position 0, not at position 1).
+	if(positions > 1)
+		positionPx /= positions - 1;
+
+	// Calculate position of the current position.
+	return startPx + positionPx * linePosition;
+}
+
+
+
+int FormationPattern::Rotatable() const
+{
+	return rotatable;
+}
+
+
+
+bool FormationPattern::FlippableY() const
+{
+	return flippable_y;
+}
+
+
+
+bool FormationPattern::FlippableX() const
+{
+	return flippable_x;
+}
+
+
+void FormationPattern::MultiAxisPoint::Add(Axis axis, const Point& toAdd)
+{
+	position[axis] += toAdd;
+}
+
+
+
+void FormationPattern::MultiAxisPoint::AddLoad(const DataNode &node)
+{
+	// We need at least the position keyword and 2 coordinate numbers.
+	if(node.Size() < 3)
+		return;
+
+	// By default we parse for pixels.
+	Axis axis = PIXELS;
+	double scalingFactor = 1.;
+
+	// Parse all the keywords before the coordinate
+	for(int i = 1; i < node.Size() - 2; ++i)
+		node.PrintTrace("Ignoring unrecognized token " + node.Token(i) + ":");
```

why are you doing this?

---

PR https://github.com/endless-sky/endless-sky/pull/4471 can parse many more tokens at this position than that this PR can. Quite a number of data-files already exist and I prefer to give an explicit warning if somebody tries to use an unsupported file, than to just silently ignore the extra tokens.

---

ah, ok